### ISSUE TITLE
Dynamic userdefined Compares and EqualityComparers with Using

### ISSUE THREAD

**Issue by NUnitTester** [nunit/nunit#2545]
if myNumbers is an array of integers, this should work:
`Assert.That(myNumbers, Is.Unique.Using((x,y) => x % 256 == y % 256));`
`Assert.That(myNumbers, Is.Unique.Using((x, y) => (x % 256).CompareTo(y % 256)));`

The only thing that works doesn't look quite fine:
`Assert.That(myNumbers, Is.Unique.Using((Comparison<int>)((x, y) => (x % 256).CompareTo(y % 256))));`

The cast is needed because of the untyped Usings:
`public CollectionItemsEqualConstraint Using(IComparer comparer);`
`public CollectionItemsEqualConstraint Using(IEqualityComparer comparer);`

My suggest is to remove the untyped Usings for compilation for .NET 4.x and to make onother overload:
`public delegate bool EqualityComparison<in T>(T x, T y);`
`public CollectionItemsEqualConstraint Using<T>(EqualityComparison<T> comparer);`

**Comment by jnm2** [comment]
I agree it's a pain. The problem is, removing those overloads will break people's code if they do anything like this:

```cs
[Test]
public void SomeTest()
{
    List<Foo> list = new List<Foo> { new Foo(x), new Foo(y) };
    IComparer someComparer = new SomeFramework.FooComparer(x, y, z);

    Assert.That(list, Is.Unique.Using(someComparer));
}
```

We just did add the `bool` delegate you're asking for in https://github.com/nunit/nunit/issues/2418, which will be released as part of NUnit 3.9 in a few days if all goes well. It's not contravariant but if that blocks someone it's easy enough to do.

**Comment by NUnitTester** [comment]
Nice to hear that the things go better with 3.9!

Its clear that's a broken change and will occure in cases of projects which are using older .NET-Frameworks more often, I think. So I proposed this change only for the compilation for newer .NET Frameworks starting with 4.0. I think there are compilations for 4.0 and 4.5 deployed.

Perhaps a wrapper class would help or a other method
`Assert.That(list, Is.Unique.UsingTypeless(someComparer));`

**Comment by CharliePoole** [comment]
There's nothing to prevent our adding typed compared overloads. The compatibility issue only comes up if we remove the untyped versions as requested.

**Comment by jnm2** [comment]
By typed, do you mean putting the type in the method name?

**Comment by jnm2** [comment]
Just so we're all on the same page, [these  are the overloads we currently have](https://github.com/nunit/nunit/blob/34026d878bbc8234b3c3efd179b0cbb9c72b4e3e/src/NUnitFramework/framework/Constraints/CollectionItemsEqualConstraint.cs#L92-L146):

```
Using(IComparer comparer)
Using<T>(IComparer<T> comparer)
Using<T>(Comparison<T> comparer)
Using(IEqualityComparer comparer)
(IEqualityComparer<T> comparer)
```

EqualConstraint also has these which CollectionItemEqualConstraint does not have:
```
Using<T>(Func<T, T, bool> comparer)
Using<TCollectionType, TMemberType>(Func<TCollectionType, TMemberType, bool> comparison)
```

Meaning I was wrong, the `bool` delegate is on `EqualConstraint` only and not on `CollectionItemsEqualConstraint`. I'd like to fix this for 3.9 if it's not too late.

**Comment by jnm2** [comment]
Gotcha. (That was not me. 😁)

We already have all the generic comparers added as overloads; there aren't any more we could add that I can think of.

Except this delegate overload. I'd like do make sure we get this: https://github.com/nunit/nunit/issues/2551

**Comment by jnm2** [comment]
Also related, me complaining about the same overload ambiguity problem with StringComparer: https://github.com/nunit/nunit/issues/2492 That one is easily solvable with the StringComparison enum but won't affect the general case.

The illusion is that `Using` is a great name because it's nice and short. The reality is that you are forced to type `.Using((IComparer<string>)`...`)`. I'd like it to be nicer.

**Comment by CharliePoole** [comment]
Does the need to cast only come up when using lambdas? I haven't run into it.

**Comment by jnm2** [comment]
The need to cast comes up for me when using comparer classes but not lambdas.

Ah, maybe @NUnitTester was expecting this to work:

```cs
Has.All.Unique.Using((a, b) => a.MyCompareTo(b));
```

But that gives you this in the error list:
> CS1660 Cannot convert lambda expression to type 'IComparer' because it is not a delegate type

Which might lead you to think you need to cast or create the delegate explicitly:
```cs
Has.All.Unique.Using(new Comparison<int>((a, b) => a.CompareTo(b)))
```

But if you hover over the first red squiggly a different message is shown:
> The type arguments for method CollectionItemsEqualConstraint.Using cannot be inferred from the usage. Try specifying the type arguments manually.

Et voila:
```cs
Has.All.Unique.Using((int a, int b) => a.CompareTo(b))
```

**Comment by NUnitTester** [comment]
Thanks, this option does work. And if the second line works in 3.9 I will be glad.
`Assert.That(myNumbers, Is.Not.Unique.Using((int x, int y) => (x % 256).CompareTo(y % 256)));`
`Assert.That(myNumbers, Is.Not.Unique.Using((int x, int y) => x % 256 == y % 256));`

**Comment by jnm2** [comment]
Cool. Let us know if you have any more questions.

**Comment by jnm2** [comment]
Update: I learned in https://github.com/nunit/nunit/issues/2552 that you will never be able to do the second line you have there:

```cs
Assert.That(myNumbers, Is.Unique.Using((int x, int y) => x % 256 == y % 256));
```

But this will work and is less to specify:

```cs
Assert.That(myNumbers, Is.Unique.Using<int>((x, y) => x % 256 == y % 256));
```
